streams/src/test/java/org/apache/kafka/streams/integration/WindowedChangelogRetentionIntegrationTest.java

```diff
@@ -0,0 +1,326 @@
+        streamsConfiguration.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
+        streamsConfiguration.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        streamsConfiguration.put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath());
+        streamsConfiguration.put(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG, 0);
+        streamsConfiguration.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, 100);
+        streamsConfiguration.put(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, Serdes.String().getClass());
+        streamsConfiguration.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serdes.Integer().getClass());
+
+        final KeyValueMapper<Integer, String, String> mapper = MockMapper.selectValueMapper();
+        final KStream<Integer, String> stream = builder.stream(STREAM_ONE_INPUT, Consumed.with(Serdes.Integer(), Serdes.String()));
+        groupedStream = stream.groupBy(mapper, Grouped.with(Serdes.String(), Serdes.String()));
+    }
+
+    @After
+    public void whenShuttingDown() throws Exception {
+        if (kafkaStreams != null) {
+            kafkaStreams.close();
+        }
+        IntegrationTestUtils.purgeLocalStreamsState(streamsConfiguration);
+    }
+
+    @Test
+    public void timeWindowedChangelogShouldHaveRetentionOfWindowSizeIfWindowSizeLargerThanDefaultRetention() throws Exception {
+        final Duration windowSize = DEFAULT_RETENTION.plus(Duration.ofHours(1));
+        final Duration expectedRetention = windowSize;
+        runAndVerifyTimeWindows(TimeWindows.of(windowSize), null, expectedRetention);
+    }
+
+    @Test
+    public void timeWindowedChangelogShouldHaveDefaultRetentionIfWindowSizeLessThanDefaultRetention() throws Exception {
+        final Duration windowSize = DEFAULT_RETENTION.minus(Duration.ofHours(1));
+        final Duration expectedRetention = DEFAULT_RETENTION;
+        runAndVerifyTimeWindows(TimeWindows.of(windowSize), null, expectedRetention);
+    }
+
+    @Test
+    public void timeWindowedChangelogShouldHaveRetentionOfWindowSizePlusGraceIfWindowSizePlusGraceLargerThanDefaultRetention() throws Exception {
+        final Duration windowSize = DEFAULT_RETENTION.plus(Duration.ofHours(1));
+        final Duration grace = Duration.ofHours(12);
+        final Duration expectedRetention = windowSize.plus(grace);
+        runAndVerifyTimeWindows(TimeWindows.of(windowSize).grace(grace), null, expectedRetention);
+    }
+
+    @Test
+    public void timeWindowedChangelogShouldHaveRetentionOfWindowSizePlusGraceIfWindowSizePlusGraceLessThanDefaultRetention() throws Exception {
+        final Duration grace = Duration.ofMillis(1000);
+        final Duration windowSize = DEFAULT_RETENTION.minus(grace).minus(Duration.ofMillis(500));
+        final Duration expectedRetention = windowSize.plus(grace);
+        runAndVerifyTimeWindows(TimeWindows.of(windowSize).grace(grace), null, expectedRetention);
```

Comment:
With 3.0 we do not use the default retention if window size + gap is less than the default retention. I think that is fine because a changelog topic that is created with 2.8 has a larger retention time than a changelog topic created with 3.0. \cc @mjsax